### ISSUE TITLE
Search improvement ideas

### ISSUE THREAD

**Issue by eddavies** [cassproject/cass-editor#286]
If a user searches: Computer Science. They will get results for everything with computer and everything for science, instead of just everything with computer science. If there are thousands and thousands of frameworks, this will make it difficult to find anything, especially since competencies within the framework will be added as results. 

How can we improve this?

**Comment by Lomilar** [comment]
Some quick facts:

1. Results are scored. Investigating why things aren't coming up in 'natural order' should be the first step.
2. Previously, search results were sorted. This is no longer the case.
3. Frameworks with competencies that match inside show up after frameworks that match. These are two separate searches, and are independent, but this isn't communicated (should put a divider between them? Search Result tabs?)

**Comment by Lomilar** [comment]
Search results seem to be accurate, but there is confusion about why some results were chosen over others. Ex: The below first result is "Framework Test case 1" but it has an exact keyword "Computer Science" -- making it a good candidate. We don't explain *why* very well.

![image](https://user-images.githubusercontent.com/337692/52222591-b75bf180-2858-11e9-9f71-0464f79038ab.png)

**Comment by Lomilar** [comment]
Another couple of ideas:

Have a search icon next to the Copy URL icon on a Creator that allows you to search for frameworks by that Creator. Would also mask the search term "<uri>" behind author:<author name>

When you begin typing a search query, autocomplete searches the registry for organizations with that partial name and gets their URI and then does something similar to the above functionally.

**Comment by Lomilar** [comment]
https://github.com/CredentialEngine/CompetencyFrameworks/issues/45

**Comment by woodkri** [comment]
Is there a way to search the registry for organizations without using an API key? Or a way for the CaSS Editor to access the API key of the organization the user is signed in under?

**Comment by cwd-mparsons** [comment]
@woodkri Yes. How would you want to use this?
There is an API to get the org by CTID and then return the apiKey and/or additional org data.

**Comment by woodkri** [comment]
@cwd-mparsons I'm working on enabling searching in the editor by organization name. To do that I need organization names and ids from the registry. I've figured out how to get those using the CE apiKey and your search API, but I'm trying to figure out how to get at that information without hardcoding an apiKey into our search. I thought one way would be to use the apiKey of the user who's logged in, if we're able to access that somehow.

We do get the CTID from the identity PostMessage CE sends, so sounds like that could work. Could you point me in the direction of getting an apiKey from the CTID?

**Comment by cwd-mparsons** [comment]
@woodkri 
The POST endpoint is: 
https://credentialengine.org/accountsApi/Organization/GetByCTID?password=[applicationPassword]&CTID=[org CTID]

The returned object is like the following:
![image](https://user-images.githubusercontent.com/26332112/60287822-75e7e380-98d8-11e9-9319-d66db50b5ad7.png)

We could set up a specific password for use by CaSS. We sometimes use a unique password for each environment (test, staging, production). For your deployment, do prefer a single password, or one per environment?

**Comment by stuartasutton** [comment]
Sorry to join this thread so late...and likely misunderstand the problem. But, has this search been set up to follow established search conventions? I.e., if I search for "computer science" with quotations marks, I am searching for the phrase and retrieve only works with the phrase computer science. If I search for computer science without quotes, the return is all works where either or both terms appear anywhere in the works.

**Comment by woodkri** [comment]
@stuartasutton In this case, we would be searching the registry for organization names that match what the user is typing in. If the user selects one of the options, we'd then query CaSS using the id of that organization which would be an exact match. I'm not sure how the registry's search is set up.

Regarding our regular framework and competency search, we use Elasticsearch's simple query string, which does search for the phrase when wrapped in quotation marks.

@cwd-mparsons Thanks for this information. I'm leaning towards getting a token from the publisher, but I'll talk it over with the team and maybe we can go over options in an upcoming meeting.

**Comment by stuartasutton** [comment]
Ah...ok

**Comment by siuc-nate** [comment]
Sample query to get organizations by name:
```
{
	"Query": {
		"@type": [
			"ceterms:CredentialOrganization",
			"ceterms:QACredentialOrganization"
		],
		"ceterms:name": "some org name"
	},
	"Skip": 0,
	"Take": 5
}
```

You can try this out here:
https://credreg.net/quickstart/queryhelper
You'll only need the inner part of the query:
```
{
	"@type": ["ceterms:CredentialOrganization", "ceterms:QACredentialOrganization"],
	"ceterms:name": "some org name"
}
```

**Comment by Lomilar** [comment]
There are two strategies, that should be chosen between based on the ceasnDataFields=true and ceApiKey=<key> both being present (for the CE method) and the more general method for everything else.

# CE Method
Get ceApiKey from URL params.
Search for all organizations at the start of the app.
Build search index / autosuggest based on Creator, Publisher.

Search index: @id <-> creator, publisher based on name

# General Method
Search for all Person objects and all Organization objects.
Build search index / autosuggest based on Owner, Reader.

Search index: @id guid <-> EcPpk.parse("@owner").fingerprint(), EcPpk.parse("@reader").fingerprint() based on name.

**Comment by woodkri** [comment]
@siuc-nate @cwd-mparsons Is CORS enabled on the search API? My search query works in Postman but I'm getting CORS errors when I try to post with the Authorization header in the browser.

I was able to get the version without the token to work through the browser but the header for authorization is triggering a preflight request.

**Comment by siuc-nate** [comment]
No, the API key is supposed to be secret - unless there's a way to send it through javascript that doesn't expose it to the client.  Usually we proxy requests through our own server for that kind of thing to other APIs we use.

**Comment by Lomilar** [comment]
Given the individual using the application would only be able to get their own API key (as it would be passed in / retrieved by the parent page or fetched via an SSO login OAuth param), I'm not sure that's sufficient reason to enforce CORS deny.

As they are the ones making the API request on behalf of their own organization, using our organization's key doesn't make sense either. 

BTW: This is the thick client argument against server based API keys. This way keeps more reliable metrics about what organizations are hitting the search API, not just which apps. It also lets you deny access at an organization level for abuse, instead of an at an app-level that happens to be proxying attacks -- which defeats the purpose of the API key.

**Comment by siuc-nate** [comment]
Right, I keep forgetting you plan to fetch/use the organization's API key. Or is that something we'd need to pass in along with the public/private key data?

**Comment by Lomilar** [comment]
That's something you'll have to pass in along with the public/private key data.
